### PR TITLE
Remove concepts of plain and simple literals

### LINKED ISSUE #26: RDF 1.1 alignment - no "plain literal"
https://w3c.github.io/sparql-query/spec/index.html#modOrderBy

Remove "A plain literal is lower than an RDF literal with type xsd:string of the same lexical form."

### PR DESCRIPTION
**Pull request by rubensworks**
Instead, we refer to the concept of simple literals in RDF 1.2, and prefer explicit mentions to xsd:string in function definitions if relevant.

This also removes any examples where xsd:string and simple literals were shown as 2 distinct cases, to avoid confusion on them being different.

Closes #26

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/sparql-query/pull/57.html" title="Last updated on Apr 28, 2023, 8:37 AM UTC (2a33051)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/sparql-query/57/47e09ed...2a33051.html" title="Last updated on Apr 28, 2023, 8:37 AM UTC (2a33051)">Diff</a>

### CODE REVIEW

File: spec/index.html [hunk 1]
```diff
@@ -3702,7 +3695,7 @@ <h3>ORDER BY</h3>
         </div>
         <p>The <a href="#op_lt">"&lt;" operator</a> (see the <a href="#OperatorMapping">Operator
             Mapping</a> and <a href="#operatorExtensibility">17.3.1 Operator Extensibility</a>) defines
-          the relative order of pairs of <code>numerics</code>, <code>simple literals</code>,
+          the relative order of pairs of <code>numerics</code>,
           <code>xsd:strings</code>, <code>xsd:booleans</code> and <code>xsd:dateTimes</code>. Pairs of
           IRIs are ordered by comparing them as <code>simple literals</code>.</p>
```
Comment (hartig) [inline spec/index.html:3700]: What to do with this one?

Comment (hartig) [reply spec/index.html:3700]:
```suggestion
          IRIs are ordered by comparing them as <a data-cite="RDF12-CONCEPTS#dfn-simple-literal">simple literals</a>.</p>
```

File: spec/index.html [hunk 2]
```diff
@@ -4841,9 +4830,7 @@ <h4>Effective Boolean Value (EBV)</h4>
             <li>If the argument is a <span class="type typedLiteral">typed literal</span> with a
               <span class="type datatype">datatype</span> of <code>xsd:boolean</code>, and it has a
               valid lexical form, the EBV is the value of that argument.</li>
-            <li>If the argument is a <span class="type plainLiteral">plain literal</span> or a
-              <span class="type typedLiteral">typed literal</span> with a <span class=
-                                                                                "type datatype">datatype</span> of <code>xsd:string</code>, the EBV is false if the
+            <li>If the argument is a <span class="type simpleLiteral">simple literal</span>, the EBV is false if the
```
Comment (hartig) [inline spec/index.html:4833]: Since simple literals are something that exists only in concrete syntaxes, I would say that they are not relevant here. So, this part about the simple literal should be removed.

However, removing the case of typed literals with datatype xsd:string is incorrect.

Hence, this line should be changed as follows.
```suggestion
            <li>If the argument is a <span class="type typedLiteral">typed literal</span> with a
              <span class="type datatype">datatype</span> of <code>xsd:string</code>, the EBV is false if the
```

File: spec/index.html [hunk 3]
```diff
@@ -6370,9 +6275,8 @@ <h5>DATATYPE</h5>
             <p>Returns the <span class="type datatypeIRI">datatype IRI</span> of a
               <code>literal</code>.</p>
             <ul>
-              <li>If the literal is a typed literal, return the datatype IRI.</li>
-              <li>If the literal is a simple literal, return <code>xsd:string</code></li>
-              <li>If the literal is literal with a language tag, return
+              <li>If the literal is a literal without language tag, return the datatype IRI.</li>
```
Comment (hartig) [inline spec/index.html:6278]:
```suggestion
              <li>If the literal is a literal without language tag, return the datatype IRI of this literal.</li>
```

File: spec/index.html [hunk 4]
```diff
@@ -6451,7 +6353,7 @@ <h5>IRI</h5>
             <p>The <code>URI</code> function is a synonym for <a href=
                                                                  "#func-iri"><code>IRI</code></a>.</p>
             <p>If the function is passed an IRI, it returns the IRI unchanged.</p>
-            <p>Passing any RDF term other than a simple literal, xsd:string or an IRI is an
+            <p>Passing any RDF term other than a simple literal or an IRI is an
```
Comment (hartig) [inline spec/index.html:6356]:
```suggestion
            <p>Passing any RDF term other than a literal with datatype <code>xsd:string</code> or an IRI is an
```

File: spec/index.html [hunk 5]
```diff
@@ -6575,8 +6475,8 @@ <h5>Strings in SPARQL Functions</h5>
               <h6>String arguments</h6>
               <p>Certain functions (e.g. <a href="#func-regex">REGEX</a>, <a href=
                                                                              "#func-strlen">STRLEN</a>, <a href="#func-contains">CONTAINS</a>) take a <code>string
-                  literal</code> as an argument and accept a simple literal, a plain literal with
-                language tag, or a literal with datatype xsd:string. They then act on the lexcial form
+                  literal</code> as an argument and accept a simple literal, or a literal with
+                language tag. They then act on the lexical form
```
Comment (hartig) [inline spec/index.html:6479]:
```suggestion
                  literal</code> as an argument and accept a simple literal,
                  a literal with datatype xsd:string, or a literal with
                language tag. They then act on the lexical form
```

File: spec/index.html [hunk 6]
```diff
@@ -6586,14 +6486,13 @@ <h6>Argument Compatibility Rules</h6>
...
                 functions raises an error.</p>
               <p>Compatibility of two arguments is defined as:</p>
               <ul>
-                <li>The arguments are simple literals or literals typed as xsd:string</li>
-                <li>The arguments are plain literals with identical language tags</li>
-                <li>The first argument is a plain literal with language tag and the second argument
-                  is a simple literal or literal typed as xsd:string</li>
+                <li>The arguments are simple literals</li>
+                <li>The arguments are literals with identical language tags</li>
+                <li>The first argument is a literal with language tag and the second argument is a simple literal</li>
```
[the remark this answers is not in the document]
Comment (hartig) [reply spec/index.html:6495]: Similarly, in the table below?

File: spec/index.html [hunk 7]
```diff
@@ -6653,20 +6532,14 @@ <h6>Argument Compatibility Rules</h6>
...
                   </tbody>
                 </table>
               </div>
             </section>
             <section id="idp1915512">
               <h6>String Literal Return Type</h6>
               <p>Functions that return a string literal do so with the string literal of the same
-                kind as the first argument (simple literal, plain literal with same language tag,
-                xsd:string). This includes <a href="#func-substr">SUBSTR</a>, <a href=
+                kind as the first argument (simple literal, literal with the same language tag). This includes <a href="#func-substr">SUBSTR</a>, <a href=
```
Comment (hartig) [inline spec/index.html:6542]: Also here, why did you remove the part about xsd:string?

File: spec/index.html [hunk 8]
```diff
@@ -6710,8 +6583,7 @@ <h5>SUBSTR</h5>
                                                                                                                                                                                                                   "type">xsd:integer</span> length)</pre>
             <p>The <code>substr</code> function corresponds to the XPath <a data-cite=
                                                                             "XPATH-FUNCTIONS#func-substring">fn:substring</a> function and returns a literal of the
-              same kind (simple literal, literal with language tag, <code>xsd:string</code> typed
-              literal) as the <code>source</code> input parameter but with a lexical form formed from
+              same kind (simple literal, literal with the same language tag) as the <code>source</code> input parameter but with a lexical form formed from
```
Comment (hartig) [inline spec/index.html:6586]: ... and here.

File: spec/index.html [hunk 9]
```diff
@@ -6947,8 +6819,8 @@ <h5>STRBEFORE</h5>
               raised.</p>
             <p>For compatible arguments, if the lexical part of the second argument occurs as a
               substring of the lexical part of the first argument, the function returns a literal of
-              the same kind as the first argument <code>arg1</code> (simple literal, plain literal same
-              language tag, xsd:string). The lexical form of the result is the substring of the lexical
+              the same kind as the first argument <code>arg1</code> (simple literal, literal with the same
+              language tag). The lexical form of the result is the substring of the lexical
```
Comment (hartig) [inline spec/index.html:6823]: And here as well.

Comment (hartig) [reply spec/index.html:6823]: And below ...

File: spec/index.html (hunk 5 above)
Comment (hartig) [reply spec/index.html:6479]: In fact, I am not even sure whether simple literals should be mentioned here. As I wrote in one of my other comments above, simple literals are something that is specific in the context of concrete syntaxes as syntactic sugar for typed literals with datatype xsd:string (and no language tag). I think that, for the definition of the functions, the syntactic sugar is irrelevant.

File: spec/index.html [hunk 10]
```diff
@@ -7113,19 +6983,19 @@ <h5>CONCAT</h5>
                     <td>"foobar"@en</td>
                   </tr>
                   <tr>
-                    <td><code>concat("foo"^^xsd:string, "bar"^^xsd:string)</code></td>
-                    <td>"foobar"^^xsd:string</td>
+                    <td><code>concat("foo"^^ex:other, "bar"^^ex:other)</code></td>
```
Comment (afs) [inline spec/index.html:6986]: `^^ex:other` is an error because it is not a string.

File: spec/index.html (hunk 5 above)
Comment (afs) [reply spec/index.html:6479]: Post first pass review: maybe we do need some language, not terminology to be able to refer to "xsd:string or rdf:langString" for use in the function definitions where that case occurs frequently.

"string literal" is the best I can come up with for now - it is a bit close to "xsd:string".

A longer phrase "string or language tagged literal" somewhat defeats the point of having such language. Maybe we just have to write "xsd:string or rdf:langString".

**Comment by afs** [comment]
@rubensworks Would you mind if we `s/plain/simple/` in the title?

File: spec/index.html (hunk 5 above)
Comment (hartig) [reply spec/index.html:6479]: Agree. I think that "string literal" is a good term for this purpose.

File: spec/index.html [hunk 11]
```diff
@@ -591,16 +588,12 @@ <h4>Terminology</h4>
               <a data-cite="RDF12-CONCEPTS#dfn-blank-node">blank node</a>
             </li>
           </ul>
-          <p>In addition, we define the following terms:</p>
+          <p>In addition, we define the following term:</p>
           <ul>
             <li>
               <a class="type" href="#defn_RDFTerm">RDF Term</a>, which includes IRIs, blank nodes and
               literals
```
Comment (TallTed) [inline spec/index.html:595]: I suppose quoted triples will be added later, when SPARQL 1.2 is brought in line with RDF 1.2 (vs the current change set which seems to bring SPARQL 1.1 in line with RDF 1.1)?

File: spec/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (TallTed) [reply spec/index.html:3700]: Maybe?
```suggestion
          IRIs are ordered by comparing their lexical rendition.</p>
```

File: spec/index.html [hunk 12]
```diff
@@ -6653,20 +6532,14 @@ <h6>Argument Compatibility Rules</h6>
...
                       <td>no</td>
                     </tr>
-                    <tr>
-                      <td>"abc"^^xsd:string</td>
-                      <td>"b"@en</td>
-                      <td>no</td>
-                    </tr>
                   </tbody>
                 </table>
               </div>
```
[the remark this answers is not in the document]
Comment (TallTed) [reply spec/index.html:6537]: For better reader clarity, I strongly urge wrapping a great many things in `<code>` tags, such as --
```suggestion
              </div>
              <p class="note">
                <code>"abc"</code> is a                 
                <a data-cite="RDF12-CONCEPTS#dfn-simple-literal">simple literal</a>
                syntactic shorthand for <code>"abc"^^xsd:string</code>.
              </p>
```

File: spec/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (hartig) [reply spec/index.html:3700]: Or the following? (assuming we define the term "string literal" as discussed somewhere else in this PR)
```suggestion
          IRIs are ordered by comparing corresponding string literals with the IRIs as their lexical form.</p>
```

Comment (afs) [reply spec/index.html:3700]: > What to do with this one?

It can be removed. It's covered by `xsd:string`.

[the remark this answers is not in the document]
Comment (afs) [reply spec/index.html:3700]:
```suggestion
          IRIs are ordered by comparing them as a sequence characters.</p>
```
this is avoiding "lexical" language which is tied closely to literals.

The URI RFC abstract says "A Uniform Resource Identifier (URI) is a compact sequence of characters that identifies an abstract or physical resource"

File: spec/index.html (hunk 5 above)
Comment (rubensworks) [reply spec/index.html:6479]: I agree with the term "string literal".

File: spec/index.html (hunk 1 above)
Comment (rubensworks) [reply spec/index.html:3700]: I'll do a full pass of the document to replace any (relevant) mentions of "simple literals".

File: spec/index.html (hunk 3 above)
Comment (rubensworks) [reply spec/index.html:6279]: > Can we just remove the whole <ul> now that all literals have datatypes?

That should be sufficient indeed. Will look into removing the two cases.

I do think it make sense to add a note, reminding the reader that language-tagged strings have a dedicated datatype.

File: spec/index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (rubensworks) [reply spec/index.html:6496]: Will use `xsd:string` instead of simple literals, as discussed above.

File: spec/index.html [hunk 13]
```diff
@@ -6608,31 +6507,11 @@ <h6>Argument Compatibility Rules</h6>
                       <td>"b"</td>
                       <td>yes</td>
                     </tr>
-                    <tr>
```
[the remark this answers is not in the document]
Comment (rubensworks) [reply spec/index.html:6611]: I removed these cases to avoid confusion, as readers might think they are different, while they are not.

> Or remove and have a respec "note" to remind the reader that "abc" is "abc"^^xsd:string.

I agree that this makes sense, will look into this.

**Comment by rubensworks** [comment]
> @rubensworks Would you mind if we s/plain/simple/ in the title?

Updated. But changed to both "plain" and "simple", since the goal is to remove both.

File: spec/index.html (hunk 11 above)
Comment (rubensworks) [reply spec/index.html:595]: Yep, that's the plan.

**Comment by rubensworks** [comment]
All comments above should now be resolved. Also removed the concept of _typed literals_, since that separate concept has also been removed now from RDF, as all literals are typed now.

**Review by TallTed — APPROVED** [review]

**Review by hartig — APPROVED** [review]

**Review by afs — APPROVED** [review]

**Comment by afs** [comment]
I've resolved the conflicts with c9c1d10.

It is not, at the moment, allowing "rebase and merge" but does allow "squash and merge".

This has happened before on another document - the block on "rebase and merge" went away (guess: eventual consistency in the GH platform).

**Comment by afs** [comment]
Looking through the doc, the only in mention of plain literal (RDF 1.0 sense) is in: "17.2.2 Effective Boolean Value (EBV)"

"If the argument is a plain literal or a typed literal with a datatype of xsd:string,..."

Fixed by fa69fbc.

I believe this PR is ready to merge and we can then prepare for FPWD.

**Comment by rubensworks** [comment]
I guess we can just squash and merge then? (feel free to do so if you agree @afs)

**Comment by afs** [comment]
Added two small fixes that I came across while checking the document.

**Review by TallTed — COMMENTED** [review]
@afs — Sadly, 24 hours from Friday ends on Saturday, so these small notes didn't get in before you merged. I can make these into a PR, or leave them with you. Let me know which you prefer.

**Comment by afs** [comment]
@TallTed - a PR please.

I don't see a way in the UI to extract them.

**Comment by gkellogg** [comment]
If the branch used for the PR is still around (or it can be restored), you can accept the changes and do a new PR from that branch, or one based on it.

**Comment by afs** [comment]
I've restored the branch but I'm not seeing a way to accept the changes because "pull request closed".

A PR would be safer.

**Comment by gkellogg** [comment]
Yeah, you would create a new PR from the "rdf11-align" branch, or you could probably create a new branch and merge in just the commits you want.

**Comment by TallTed** [comment]
I'll make a PR against `main` later today.

**Comment by TallTed** [comment]
See https://github.com/w3c/sparql-query/pull/62
